src/serde.rs

```diff
@@ -0,0 +1,201 @@
+extern crate std;
+use std::prelude::v1::*;
+
+use {
+    super::{StrIndex, StrRange},
+    serde::{
+        de::{Deserialize, Deserializer, Error, MapAccess, SeqAccess, Visitor},
+        ser::{Serialize, SerializeStruct, Serializer},
+    },
+    std::fmt,
+};
+
+impl Serialize for StrIndex {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_newtype_struct("StrIndex", &self.raw)
+    }
+}
+
+impl<'de> Deserialize<'de> for StrIndex {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_newtype_struct("StrIndex", StrIndexVisitor)
+    }
+}
+
+impl Serialize for StrRange {
```

Comment:
Hm, maybe serialize this as newtype warapper of a tupple? Should make JSON more compact and readable.

Reply:
Personally, I think I slightly prefer `{ "start": 0, "end": 10 }` to `[0, 10]` (especially since that looks like `[0..=10]` rather than `[0..10)`!), though I understand that requires manually going and adjusting the pretty JSON formatting.

Also, mostly note to self but also readers, the serialization format does not mean it has to be represented in memory that way. We could switch to `{ start: u32, len: u16 }` and still serialize as `{ start: u32, end: u32 }` just fine, if we so desired.